src/server.c

```diff
@@ -4997,6 +5001,88 @@ void linuxMemoryWarnings(void) {
         serverLog(LL_WARNING,"WARNING you have Transparent Huge Pages (THP) support enabled in your kernel. This will create latency and memory usage issues with Redis. To fix this issue run the command 'echo madvise > /sys/kernel/mm/transparent_hugepage/enabled' as root, and add it to your /etc/rc.local in order to retain the setting after a reboot. Redis must be restarted after THP is disabled (set to 'madvise' or 'never').");
     }
 }
+
+static int smapsGetSharedDirty(pid_t pid, unsigned long addr) {
+    int ret, in_mapping = 0, val = -1;
+    unsigned long from, to;
+    char buf[64];
+    FILE *f;
+
+    ret = snprintf(buf, sizeof(buf), "/proc/%d/smaps", pid);
+    serverAssert((unsigned int)ret < sizeof(buf));
+
+    f = fopen(buf, "r");
+    serverAssert(f);
+
+    while (1) {
+        if (!fgets(buf, sizeof(buf), f))
+            break;
+
+        ret = sscanf(buf, "%lx-%lx", &from, &to);
+        if (ret == 2)
+            in_mapping = from <= addr && addr < to;
+
+        if (in_mapping && !memcmp(buf, "Shared_Dirty:", 13)) {
+            ret = sscanf(buf, "%*s %d", &val);
+            serverAssert(ret == 1);
+            break;
+        }
+    }
+
+    fclose(f);
+    return val;
+}
+
+void linuxMadvFreeForkBugCheck(void) {
```

Comment:
please add a big comment explaining:
1. what it does (i.e. check for kernel COW bug, basically your commit comment).
2. why we do it (i.e. we want to warn the user and exit)
3. how we do it (what are the steps, or alternatively just add a few comments inside the function)